announce.py

```diff
@@ -0,0 +1,99 @@
+﻿import version
+
+class IAnnouncerDb:
+	def has_guild_seen_version(self, guild, version):
+		pass
+
+class IAnnouncerGuildDb:
+	def get_guilds():
+		pass
+
+class IAnnouncerMessageSender:
+	def send_embed(self, guild, embed):
+		pass
+
+class AnnouncerDbImpl(IAnnouncerDb):
+	def __init__(self, mongo):
+		self.collection = mongo['GuildVersionAnnouncements']
+
+	def has_guild_seen_version(self, guild, version):
+		query = { "guild" : guild }
+		doc = self.collection.find_one(query)
+		if doc:
+			vers = tuple(doc['version'])
+			return (version <= vers)
+		else:
+			return False
+
+	def record_guild_seen_version(self, guild, version):
+		query = {"guild" : guild }
+		doc = self.collection.find_one(query)
+		if doc:
+			doc['version'] = list(version)
+		else:
+			doc = {"guild" : guild, "version" : list(version)}
+		self.collection.replace_one(query, doc, True)
+
+
+class AnnouncerGuildDbImpl(IAnnouncerGuildDb):
+	def __init__(self, mongo):
+		self.collection = mongo['GuildInfo']
+
+	def get_guilds(self):
+		x = self.collection.find(projection={'guild':True, '_id':False}).distinct('guild')
+		return x
+
+class AnnouncerMessageSenderImpl(IAnnouncerMessageSender):
+	def __init__(self, bot, mongo):
+		self.bot = bot
+		self.collection = mongo['GuildInfo']
+
+	async def send_embed(self, guild, embed):
+		query = {"guild": guild}
+		result = self.collection.find(query)
+		for x in result:
+			townInfo = self.bot.getTownInfoByIds(guild, x['controlChannelId'])
+			await townInfo.controlChannel.send(embed=embed)
+
+class AnnouncerImpl:
+	def __init__(self, db, guildDb, provider, sender):
+		self.db = db
+		self.guildDb = guildDb
+		self.provider = provider
+		self.sender = sender
+
+	def guild_no_announce(self, guild):
+		hugeCrazyVersion = (999,0,0)
+		self.db.record_guild_seen_version(guild, hugeCrazyVersion)
+
+	async def announce_latest_version(self):
+		guilds = self.guildDb.get_guilds()
+		numSent = 0
+
+		if len(guilds) > 0:
+			versions = self.provider.get_versions_and_embeds()
+
+			for g in guilds:
+				for (v, embed) in versions.items():
+					if not self.db.has_guild_seen_version(g, v):
+						self.db.record_guild_seen_version(g, v)
+						numSent += 1
+						await self.sender.send_embed(g, embed)
```

Comment:
Put in a test in a case where send_embed throws an exception, announce_latest_version should continue without excepting.

This seems like a likely scenario if a guild gets deleted or permissions are removed.